### DISCUSSION TITLE
Having multiple lifespan contexts for an app.

### DISCUSSION THREAD

**Discussion by farzbood** [fastapi/fastapi#10083] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
from redis.asyncio import Redis
from sqlalchemy.ext.asyncio import create_async_engine

app = FastAPI()

@app.lifespan
async def setup_db(app: FastAPI) -> AsyncIterator[State]:
    engine = await create_async_engine("postgresql+asyncpg://user:password@localhost/db")

    yield {"db": engine}

    await engine.dispose()


@app.lifespan
async def setup_cache(app: FastAPI) -> AsyncIterator[State]:
    redis = await Redis.from_url("redis://localhost:6379/0")

    yield {"cache": redis}

    await redis.close()
    await redis.wait_closed()

@app.get("/")
def home(req: Request):
    # do something, while having access to both "db" and "cache" in <req.state>
    ...
```

### Description

Trying to separate different resource-handlers based on their domain/context and having them globally accessible, expecting to be able to define and add more than one <**lifespan**> **_AsyncContextManager_** for an **FastaAPI** app.
Current behavior is, just one **_AsyncCoroutineFunction_** can define and set as <**lifespan**> for an app.

### Operating System

Windows

### FastAPI Version

0.101.1

### Pydantic Version

2.1.1

### Python Version

3.11.2

### Additional Context

Here is an implementation of a <**LifespanManager**> that works perfectly for that purpose.
https://github.com/uriyyo/fastapi-lifespan-manager/tree/main

P.S
Such feature becomes very handy and even crucial in cases that the app developers want to have more control on the resources among separate use-cases (better **_SoC_**) and/or when **3rd-party** **_libraries/plugins/modules_** needs to do some processes such as **_initialisations_** before the app starts and some _**clean-ups**_ after the app shuts down.

e.g: https://github.com/tortoise/tortoise-orm/blob/0cb82bb4bf40bd2c2257b8dc572c19c0d6ff3111/tortoise/contrib/fastapi/__init__.py#L92
(This is the approach that **TOrtoiseORM** took, by adding-up their flow to on_event() handlers, which is not applicable if <**lifespan**> been set for **FastAPI** app beforehand, because it's not possible to have both simultaneously, also there's no intuitive way to access to the app's pre-start stage!)

Keep Up With Outstanding Work
PEACE V

**Comment by farzbood** [comment]
Hello @tiangolo  
Since this situation will soon come across many of **FastAPI** users regarding integration with 3rd-party apps/libraries relying on **startup/shutdown events** (announced deprecated and will remove) and there's no answers recieved yet, would you kindly consider this, in case?

Thank you in advance
Regards

**Reply by bartpijn** [reply]
I am currently running into this issue while building fastapi libraries to provide boilerplate code for all my FastAPI projects.

**Comment by tomschr** [comment]
Thanks for this great project! :heart:

I have exactly the same problem and found this discussion after searching for a solution. What I'm puzzled that the `lifespan` doesn't provide a sequence like the `on_startup`/`on_shutdown` events:

* `on_startup`/`on_shutdown`: type is `Optional[Sequence[Callable[[], Any]]]`
* `lifespan`: type is `Optional[Lifespan[Any]]`

As shown by @farzbood it would make so much more sense. 

It would be very helpful if we could have a sequence of lifespan async context managers to separate different logic (as shown above). :slightly_smiling_face:
